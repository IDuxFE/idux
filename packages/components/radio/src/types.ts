/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { AbstractControl } from '@idux/cdk/forms'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { SpaceProps } from '@idux/components/space'
import type { DefineComponent, HTMLAttributes, LabelHTMLAttributes, PropType } from 'vue'

export const radioProps = {
  control: { type: [String, Number, Object] as PropType<string | number | AbstractControl>, default: undefined },
  checked: { type: Boolean, default: undefined },

  autofocus: { type: Boolean, default: false },
  buttoned: { type: Boolean, default: undefined },
  disabled: { type: Boolean, default: undefined },
  label: { type: String, default: undefined },
  mode: { type: String as PropType<RadioMode>, default: undefined },
  size: { type: String as PropType<FormSize>, default: undefined },
  value: { type: null, default: undefined },
  waveless: { type: Boolean, default: false },

  // events
  'onUpdate:checked': { type: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>> },
  onChange: { type: [Function, Array] as PropType<MaybeArray<(checked: boolean, oldChecked: boolean) => void>> },
  onBlur: { type: [Function, Array] as PropType<MaybeArray<(evt: FocusEvent) => void>> },
  onFocus: { type: [Function, Array] as PropType<MaybeArray<(evt: FocusEvent) => void>> },
} as const

export type RadioProps = ExtractInnerPropTypes<typeof radioProps>
export type RadioPublicProps = ExtractPublicPropTypes<typeof radioProps>
export interface RadioBindings {
  blur: () => void
  focus: (options?: FocusOptions) => void
}
export type RadioComponent = DefineComponent<
  Omit<LabelHTMLAttributes, keyof RadioPublicProps> & RadioPublicProps,
  RadioBindings
>
export type RadioInstance = InstanceType<DefineComponent<RadioProps, RadioBindings>>

export const radioGroupProps = {
  control: { type: [String, Number, Object] as PropType<string | number | AbstractControl>, default: undefined },
  value: { type: null, default: undefined },

  buttoned: { type: Boolean, default: false },
  dataSource: { type: Array as PropType<RadioData[]> },
  disabled: { type: Boolean, default: false },
  gap: { type: [Number, String] as PropType<number | string>, default: undefined },
  name: { type: String, default: undefined },
  mode: { type: String as PropType<RadioMode>, default: undefined },
  size: { type: String as PropType<FormSize>, default: 'md' },

  // events
  'onUpdate:value': { type: [Function, Array] as PropType<MaybeArray<(value: any) => void>> },
  onChange: { type: [Function, Array] as PropType<MaybeArray<(value: any, oldValue: any) => void>> },
} as const

export type RadioGroupProps = ExtractInnerPropTypes<typeof radioGroupProps>
export type RadioGroupPublicProps = ExtractPublicPropTypes<typeof radioGroupProps> & Omit<SpaceProps, 'size'>
export type RadioGroupComponent = DefineComponent<
  Omit<HTMLAttributes, keyof RadioGroupPublicProps> & RadioGroupPublicProps
>
export type RadioGroupInstance = InstanceType<DefineComponent<RadioGroupProps>>

export type RadioMode = 'default' | 'primary'
export interface RadioData<K = VKey> extends Omit<RadioPublicProps, 'value'> {
  key?: K
}
