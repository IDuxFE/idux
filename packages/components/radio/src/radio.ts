import { computed, ComputedRef, inject } from 'vue'
import { RadioProps, RadioGroupConfig, radioModeConfig } from './types'
import type { InjectionKey } from 'vue'
import isEmpty from 'lodash/isEmpty'
export const getRadioAttrs = (
  props: RadioProps,
  isGroup: ComputedRef<boolean>,
  radioGroup: RadioGroupConfig,
): Record<string, ComputedRef> => {
  const isDisabled = computed(() => {
    return isGroup.value ? radioGroup.disabled : props.disabled
  })
  return {
    isDisabled,
  }
}
/* eslint-disable @typescript-eslint/no-explicit-any */
export const radioGroupKey: InjectionKey<RadioGroupConfig> = 'IxRadioGroup' as any

export const radioMode = (): radioModeConfig => {
  const radioGroup = inject(radioGroupKey, {} as RadioGroupConfig)
  const isGroup = computed(() => !isEmpty(radioGroup))
  return {
    isGroup,
    radioGroup,
  }
}
