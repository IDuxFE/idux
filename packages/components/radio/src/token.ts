/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { RadioGroupProps } from './types'
import type { ValueAccessor } from '@idux/cdk/forms'
import type { InjectionKey } from 'vue'

export interface RadioGroupContext {
  props: RadioGroupProps
  accessor: ValueAccessor
}

export const radioGroupToken: InjectionKey<RadioGroupContext> = Symbol('radioGroupToken')
