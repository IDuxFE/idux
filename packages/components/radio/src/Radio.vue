<template>
  <label
    :class="{
      'ix-radio': true,
      'ix-radio-disabled': isDisabled,
      'ix-radio-checked': isChecked,
    }"
    role="radio"
    :ariaChecked="isChecked"
    :ariaDisabled="isDisabled"
    @change="isChecked = true"
  >
    <span
      class="ix-radio-body"
      :class="{
        'ix-radio-body-disabled': isDisabled,
        'ix-radio-body-checked': isChecked,
      }"
    >
      <span class="ix-radio-body-inner">
        <input
          ref="radioRef"
          type="radio"
          :name="name"
          :checked="isChecked"
          ariaHidden="true"
          :disabled="isDisabled"
          class="ix-radio-body-input"
          :value="value"
          @change="onChange"
        />
      </span>
    </span>
    <span class="ix-radio-label">
      <slot></slot>
    </span>
  </label>
</template>
<script lang="ts">
import { computed, defineComponent, ref } from 'vue'
import { RadioProps, modelValueType } from './types'
import { PropTypes } from '@idux/cdk/utils'
import { getRadioAttrs, radioMode } from './radio'

export default defineComponent({
  name: 'IxRadio',
  props: {
    value: PropTypes.oneOfType([String, Number, Boolean]),
    disabled: PropTypes.bool.def(false),
    name: PropTypes.string.def(''),
    checked: PropTypes.bool.def(false),
  },
  emits: ['change', 'update:checked'],
  setup(props: RadioProps, { emit }) {
    const { isGroup, radioGroup } = radioMode()
    const { isDisabled } = getRadioAttrs(props, isGroup, radioGroup)
    const radioRef = ref<HTMLInputElement>()
    const isChecked = computed<boolean>({
      get() {
        return (isGroup.value ? radioGroup.value === props.value : props.checked) as boolean
      },
      set(value) {
        if (isGroup.value) {
          radioGroup.change(props.value as modelValueType)
        } else {
          emit('update:checked', value)
        }
        radioRef.value!.checked = value
      },
    })

    const onChange = function () {
      emit('change', isChecked)
    }

    return {
      isDisabled,
      isChecked,
      onChange,
      radioRef,
    }
  },
})
</script>
