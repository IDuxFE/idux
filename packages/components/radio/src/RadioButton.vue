<template>
  <label
    class="ix-radio-button"
    :class="{
      'ix-radio-button-checked': isChecked,
      'ix-radio-button-disabled': isDisabled,
      [`ix-radio-button-${size}`]: true,
      [`ix-radio-button-${mode}`]: true,
    }"
    role="radio"
    :ariaChecked="isChecked"
    :ariaDisabled="isDisabled"
  >
    <span class="ix-radio-button-body">
      <input
        ref="radioRef"
        type="radio"
        :name="name"
        :disabled="isDisabled"
        ariaHidden="true"
        class="ix-radio-button-body-input"
        :value="value"
        :checked="isChecked"
        @change="onChange"
      />
      <span class="ix-radio-button-inner"></span>
    </span>
    <span class="ix-radio-button-label"><slot></slot></span>
  </label>
</template>
<script lang="ts">
import { defineComponent, computed } from 'vue'
import { useGlobalConfig } from '@idux/components/config'
import { radioButtonProps } from './types'
import { getRadioAttrs, radioMode } from './radio'

export default defineComponent({
  name: 'IxRadioButton',
  props: radioButtonProps,
  emits: ['change', 'update:checked'],
  setup(props, { emit }) {
    const { isGroup, radioGroup } = radioMode()
    const { isDisabled } = getRadioAttrs(props, isGroup, radioGroup)
    const isChecked = computed<boolean>(() => {
      return (isGroup.value ? radioGroup.value === props.value : props.checked) as boolean
    })
    const onChange = (e: { target: { checked: boolean } }) => {
      const isChecked = e.target.checked
      emit('change', isChecked)
      emit('update:checked', isChecked)

      isGroup.value && radioGroup.change(props.value!)
    }
    const radioGroupConfig = useGlobalConfig('radioGroup')
    const size = computed(() => {
      return isGroup.value ? radioGroup.size : radioGroupConfig.size
    })
    const mode = computed(() => {
      return isGroup.value ? radioGroup.mode : radioGroupConfig.mode
    })

    return {
      isDisabled,
      isChecked,
      onChange,
      size,
      mode,
    }
  },
})
</script>
