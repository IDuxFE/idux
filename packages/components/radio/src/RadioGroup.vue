<template>
  <div ref="radioGroupRef" class="ix-radio-group">
    <slot></slot>
  </div>
</template>
<script lang="ts">
import { PropTypes } from '@idux/cdk/utils'
import { defineComponent, provide, reactive, toRefs } from 'vue'
import { RadioGroupProps, modelValueType } from './types'
import { radioGroupKey } from './radio'

export default defineComponent({
  name: 'IxRadioGroup',
  props: {
    value: PropTypes.oneOfType([String, Number, Boolean]),
    size: PropTypes.oneOf(['large', 'medium', 'small']).def('medium'),
    mode: PropTypes.oneOf(['fill', 'border']).def('border'),
    disabled: PropTypes.bool.def(false),
    color: PropTypes.string.def('#00b27a'),
  },
  emits: ['change', 'update:value'],
  setup(props: RadioGroupProps, { emit }) {
    const change = (radioValue: modelValueType) => {
      console.log(radioValue)
      emit('update:value', radioValue)
      emit('change', radioValue)
    }
    provide(radioGroupKey, reactive({ ...toRefs(props), change }))
  },
})
</script>
