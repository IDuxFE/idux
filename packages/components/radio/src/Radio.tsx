/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable vue/no-ref-as-operand */

import { type ComputedRef, type Ref, computed, defineComponent, inject, normalizeClass, ref } from 'vue'

import { isNil } from 'lodash-es'

import { useAccessorAndControl } from '@idux/cdk/forms'
import { callEmit } from '@idux/cdk/utils'
import { ɵWave, type ɵWaveInstance } from '@idux/components/_private/wave'
import { useGlobalConfig } from '@idux/components/config'
import { FORM_TOKEN, useFormElement, useFormItemRegister } from '@idux/components/form'
import { convertStringVNode, useKey } from '@idux/components/utils'

import { type RadioGroupContext, radioGroupToken } from './token'
import { type RadioProps, radioProps } from './types'

const buttonSizeMap = {
  sm: 'xs',
  md: 'sm',
  lg: 'md',
}

export default defineComponent({
  name: 'IxRadio',
  inheritAttrs: false,
  props: radioProps,
  setup(props, { attrs, expose, slots }) {
    const key = useKey()
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-radio`)
    const config = useGlobalConfig('radio')

    const { elementRef, focus, blur } = useFormElement<HTMLInputElement>()
    const waveRef = ref<ɵWaveInstance>()
    expose({ focus, blur })

    const formContext = inject(FORM_TOKEN, null)
    const radioGroup = inject(radioGroupToken, null)
    const mergedName = computed(() => (attrs.name as string) ?? radioGroup?.props.name)
    const mergedValue = computed(() => {
      const { value } = props
      if (!isNil(value)) {
        return value
      }
      // 当在 group 中使用时，不传 value 就使用 key 作为 value
      return radioGroup ? key : undefined
    })
    const isButtoned = computed(() => props.buttoned ?? radioGroup?.props.buttoned)
    const size = computed(() => props.size ?? radioGroup?.props.size ?? formContext?.size.value ?? config.size)
    const mergedWaveless = computed(() => props.waveless ?? config.waveless)
    const mode = computed(() => props.mode ?? radioGroup?.props.mode ?? 'default')
    const { isChecked, isDisabled, isFocused, handleChange, handleBlur, handleFocus } = useRadio(
      props,
      radioGroup,
      elementRef,
      mergedValue,
      waveRef,
      mergedWaveless,
    )
    const classes = computed(() => {
      const buttoned = isButtoned.value
      const checked = isChecked.value
      const isPrimary = buttoned && checked && mode.value === 'primary'
      const prefixCls = mergedPrefixCls.value
      const commonPrefixCls = common.prefixCls
      const classes = {
        [prefixCls]: true,
        [`${prefixCls}-checked`]: isChecked.value,
        [`${prefixCls}-disabled`]: isDisabled.value,
        [`${prefixCls}-focused`]: isFocused.value,
        [`${prefixCls}-${size.value}`]: buttoned,
        [`${commonPrefixCls}-button`]: buttoned,
        [`${commonPrefixCls}-button-default`]: buttoned && !isPrimary,
        [`${commonPrefixCls}-button-primary`]: isPrimary,
        [`${commonPrefixCls}-button-disabled`]: buttoned && isDisabled.value,
        [`${commonPrefixCls}-button-${buttonSizeMap[size.value]}`]: buttoned,
      }
      return normalizeClass([classes, attrs.class])
    })

    return () => {
      const { autofocus, label } = props
      const { class: className, style, type, tabindex, ...restAttrs } = attrs
      const prefixCls = mergedPrefixCls.value
      const labelNode = convertStringVNode(slots.default, label)
      return (
        <label
          class={classes.value}
          style={style as string}
          role="radio"
          aria-checked={isChecked.value}
          aria-disabled={isDisabled.value}
        >
          <span class={`${prefixCls}-input`}>
            <input
              ref={elementRef}
              type="radio"
              class={`${prefixCls}-input-inner`}
              aria-hidden
              autofocus={autofocus}
              checked={isChecked.value}
              disabled={isDisabled.value}
              name={mergedName.value}
              value={mergedValue.value}
              onChange={handleChange}
              onBlur={handleBlur}
              onFocus={handleFocus}
              {...restAttrs}
            />
            {!isButtoned.value && (
              <span class={`${prefixCls}-input-box`} tabindex={tabindex as number}>
                {!mergedWaveless.value && <ɵWave ref={waveRef} />}
              </span>
            )}
          </span>
          {isButtoned.value && !mergedWaveless.value && <ɵWave ref={waveRef} />}
          {labelNode && <span class={`${prefixCls}-label`}>{labelNode}</span>}
        </label>
      )
    }
  },
})

const useRadio = (
  props: RadioProps,
  radioGroup: RadioGroupContext | null,
  elementRef: Ref<HTMLInputElement | undefined>,
  mergedValue: ComputedRef<unknown>,
  waveRef: Ref<ɵWaveInstance | undefined>,
  mergedWaveless: ComputedRef<boolean>,
) => {
  let isChecked: ComputedRef<boolean>
  let isDisabled: ComputedRef<boolean>
  const isFocused = ref(false)

  let handleChange: (evt: Event) => void
  let handleBlur: (evt: FocusEvent) => void
  const handleFocus = (evt: FocusEvent) => {
    isFocused.value = true
    callEmit(props.onFocus, evt)
  }

  if (radioGroup) {
    const { accessor, props: groupProps } = radioGroup
    isChecked = computed(() => accessor.value === mergedValue.value)
    isDisabled = computed(() => accessor.disabled || !!props.disabled)
    handleBlur = (evt: FocusEvent) => {
      isFocused.value = false
      accessor.markAsBlurred()
      callEmit(props.onBlur, evt)
    }
    handleChange = (evt: Event) => {
      if (elementRef.value) {
        const checked = (evt.target as HTMLInputElement).checked
        const value = mergedValue.value
        const oldValue = accessor.value
        accessor.setValue(value)
        // 为了保持受控模式下保持原生input状态和数据一致
        elementRef.value.checked = false
        callEmit(props.onChange, checked, !checked)
        callEmit(groupProps.onChange, value, oldValue)
        !mergedWaveless.value && waveRef.value?.play()
      }
    }
  } else {
    const { accessor, control } = useAccessorAndControl({ valueKey: 'checked' })
    useFormItemRegister(control)
    isChecked = computed(() => !!accessor.value)
    isDisabled = computed(() => accessor.disabled)
    handleBlur = (evt: FocusEvent) => {
      isFocused.value = false
      accessor.markAsBlurred()
      callEmit(props.onBlur, evt)
    }
    handleChange = (evt: Event) => {
      if (elementRef.value) {
        const checked = (evt.target as HTMLInputElement).checked
        accessor.setValue(checked)
        elementRef.value.checked = false
        callEmit(props.onChange, checked, !checked)
        !mergedWaveless.value && waveRef.value?.play()
      }
    }
  }

  return { isChecked, isDisabled, isFocused, handleChange, handleBlur, handleFocus }
}
