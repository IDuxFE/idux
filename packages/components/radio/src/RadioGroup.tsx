/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { type VNodeChild, computed, defineComponent, normalizeClass, provide } from 'vue'

import { isNil } from 'lodash-es'

import { Logger, convertCssPixel } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { useFormAccessor } from '@idux/components/form'

import Radio from './Radio'
import { radioGroupToken } from './token'
import { radioGroupProps } from './types'

export default defineComponent({
  name: 'IxRadioGroup',
  props: radioGroupProps,
  setup(props, { slots }) {
    const accessor = useFormAccessor()

    provide(radioGroupToken, { props, accessor })

    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-radio-group`)

    const classes = computed(() => {
      const { gap } = props
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-with-gap`]: !isNil(gap),
      })
    })

    const style = computed(() => {
      const { gap } = props
      return gap != null ? `gap: ${convertCssPixel(gap)};` : undefined
    })

    return () => {
      const { options, dataSource } = props
      if (options) {
        Logger.warn('components/radio', '`options` was deprecated, please use `dataSource` instead')
      }
      const data = options ?? dataSource
      let children: VNodeChild[] | undefined
      if (data) {
        children = data.map(item => {
          const { key, value } = item
          return <Radio {...item} key={key ?? value} value={value ?? key} />
        })
      } else {
        children = slots.default ? slots.default() : undefined
      }
      return (
        <div class={classes.value} style={style.value}>
          {children}
        </div>
      )
    }
  },
})
