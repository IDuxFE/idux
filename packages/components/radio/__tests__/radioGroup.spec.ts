import { MountingOptions, flushPromises, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import Radio from '../src/Radio'
import RadioGroup from '../src/RadioGroup'
import { RadioGroupProps } from '../src/types'

describe('RadioGroup', () => {
  const defaultDataSource = [
    { label: 'Beijing', value: 'a' },
    { label: 'Shanghai', value: 'b' },
    { label: 'Guangzhou', value: 'c' },
    { label: 'Shenzhen', value: 'd' },
  ]
  const RadioGroupMount = (groupOptions?: MountingOptions<Partial<RadioGroupProps>>) => {
    const { props, ...rest } = groupOptions || {}
    return mount(RadioGroup, { props: { dataSource: defaultDataSource, ...props }, ...rest })
  }

  renderWork<RadioGroupProps>(RadioGroup, { props: { dataSource: defaultDataSource } })

  test('value work', async () => {
    const wrapper = RadioGroupMount({ props: { value: 'a' } })

    expect(wrapper.find('.ix-radio-checked').text()).toBe('Beijing')

    await wrapper.setProps({ value: 'b' })

    expect(wrapper.find('.ix-radio-checked').text()).toBe('Shanghai')
  })

  test('onUpdate:value work', async () => {
    const onUpdate = jest.fn()
    const wrapper = RadioGroupMount({ props: { 'onUpdate:value': onUpdate } })

    await wrapper.findAll('input')[0].setValue(true)
    await flushPromises()

    expect(onUpdate).toBeCalledWith('a')

    await wrapper.findAll('input')[1].setValue(true)

    expect(onUpdate).toBeCalledWith('b')
  })

  test('buttoned work', async () => {
    const wrapper = RadioGroupMount({ props: { buttoned: true } })

    expect(wrapper.findAll('.ix-radio-button').length).toBe(4)

    await wrapper.setProps({ buttoned: false })

    expect(wrapper.findAll('.ix-radio-button').length).toBe(0)
  })

  test('disabled work', async () => {
    const wrapper = RadioGroupMount({ props: { disabled: true } })

    expect(wrapper.findAll('.ix-radio-disabled').length).toBe(4)

    await wrapper.setProps({ disabled: false })

    expect(wrapper.findAll('.ix-radio-disabled').length).toBe(0)
  })

  test('gap work', async () => {
    const wrapper = RadioGroupMount()

    expect(wrapper.classes()).not.toContain('ix-radio-group-with-gap')

    await wrapper.setProps({ gap: 8 })

    expect(wrapper.classes()).toContain('ix-radio-group-with-gap')
    expect((wrapper.element as HTMLElement).style.gap).toEqual('8px')
  })

  test('name work', async () => {
    const wrapper = RadioGroupMount({ props: { name: 'city' } })

    expect(wrapper.findAll('input[name="city"]').length).toBe(4)

    await wrapper.setProps({ name: 'city1' })

    expect(wrapper.findAll('input[name="city1"]').length).toBe(4)
  })

  test('mode work', async () => {
    const wrapper = RadioGroupMount({ props: { mode: 'primary', buttoned: true } })

    expect(wrapper.findAll('.ix-radio-primary').length).toBe(4)

    await wrapper.setProps({ mode: 'default' })

    expect(wrapper.findAll('.ix-radio-default').length).toBe(4)
  })

  test('dataSource work', async () => {
    let dataSource = [
      { label: 'Beijing', value: 'a', disabled: true },
      { label: 'Shanghai', value: 'b' },
    ]
    const wrapper = RadioGroupMount({ props: { dataSource } })

    expect(wrapper.findAll('.ix-radio-disabled').length).toBe(1)

    expect(wrapper.findAll('.ix-radio').length).toBe(2)

    dataSource = [
      { label: 'Beijing', value: 'a' },
      { label: 'Shanghai', value: 'b' },
      { label: 'Guangzhou', value: 'c' },
    ]

    await wrapper.setProps({ dataSource })

    expect(wrapper.findAll('.ix-radio').length).toBe(3)
  })

  test('default slot work', async () => {
    const dataSource = undefined
    const slots = [h(Radio, { label: 'A' }), h(Radio, { label: 'B' }), h(Radio, { label: 'C' })]
    const wrapper = RadioGroupMount({ props: { dataSource }, slots: { default: () => slots } })

    expect(wrapper.findAll('.ix-radio').length).toBe(3)
  })

  test('size work', async () => {
    const wrapper = RadioGroupMount({ props: { size: 'lg', buttoned: true } })

    expect(wrapper.findAll('.ix-radio-lg').length).toBe(4)

    await wrapper.setProps({ size: 'sm' })

    expect(wrapper.findAll('.ix-radio-sm').length).toBe(4)

    await wrapper.setProps({ size: undefined })

    expect(wrapper.findAll('.ix-radio-md').length).toBe(4)
  })
})
