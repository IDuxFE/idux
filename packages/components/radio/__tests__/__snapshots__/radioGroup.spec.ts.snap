// Vitest Snapshot v1

exports[`RadioGroup > render work 1`] = `
"<div class=\\"ix-space ix-radio-group\\" style=\\"margin-bottom: -8px;\\">
  <div class=\\"ix-space-item\\" style=\\"margin-right: 8px; padding-bottom: 8px;\\"><label class=\\"ix-radio\\" role=\\"radio\\" aria-checked=\\"false\\" aria-disabled=\\"false\\"><span class=\\"ix-radio-input\\"><input type=\\"radio\\" class=\\"ix-radio-input-inner\\" aria-hidden=\\"true\\" value=\\"a\\"><span class=\\"ix-radio-input-box\\"><div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div></span></span>
      <!----><span class=\\"ix-radio-label\\">Beijing</span>
    </label></div>
  <div class=\\"ix-space-item\\" style=\\"margin-right: 8px; padding-bottom: 8px;\\"><label class=\\"ix-radio\\" role=\\"radio\\" aria-checked=\\"false\\" aria-disabled=\\"false\\"><span class=\\"ix-radio-input\\"><input type=\\"radio\\" class=\\"ix-radio-input-inner\\" aria-hidden=\\"true\\" value=\\"b\\"><span class=\\"ix-radio-input-box\\"><div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div></span></span>
      <!----><span class=\\"ix-radio-label\\">Shanghai</span>
    </label></div>
  <div class=\\"ix-space-item\\" style=\\"margin-right: 8px; padding-bottom: 8px;\\"><label class=\\"ix-radio\\" role=\\"radio\\" aria-checked=\\"false\\" aria-disabled=\\"false\\"><span class=\\"ix-radio-input\\"><input type=\\"radio\\" class=\\"ix-radio-input-inner\\" aria-hidden=\\"true\\" value=\\"c\\"><span class=\\"ix-radio-input-box\\"><div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div></span></span>
      <!----><span class=\\"ix-radio-label\\">Guangzhou</span>
    </label></div>
  <div class=\\"ix-space-item\\" style=\\"padding-bottom: 8px;\\"><label class=\\"ix-radio\\" role=\\"radio\\" aria-checked=\\"false\\" aria-disabled=\\"false\\"><span class=\\"ix-radio-input\\"><input type=\\"radio\\" class=\\"ix-radio-input-inner\\" aria-hidden=\\"true\\" value=\\"d\\"><span class=\\"ix-radio-input-box\\"><div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div></span></span>
      <!----><span class=\\"ix-radio-label\\">Shenzhen</span>
    </label></div>
</div>"
`;
